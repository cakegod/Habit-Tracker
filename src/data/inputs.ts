import type { UNITS } from "@util/constants";

// HELPER

export class InputBase {
  name: string;
  label: string;
  value: string;
  type: "number" | "text";
  options: [string | number, string][];
  selectedOption: keyof typeof UNITS | number;
  inputCategory: InputCategory;

  constructor({
    name,
    label,
    value = "",
    type = "number",
    options,
    selectedOption,
    inputCategory = "inputSelect",
  }: {
    name: string;
    label: string;
    value?: string;
    type?: "number" | "text";
    options: [string | number, string][];
    selectedOption: keyof typeof UNITS | number;
    inputCategory?: InputCategory;
  }) {
    this.name = name;
    this.label = label;
    this.value = value;
    this.type = type;
    this.options = options;
    this.selectedOption = selectedOption;
    this.inputCategory = inputCategory;
  }

  // get unit() {
  //   return this.selectedOption;
  // }
}

export class InputGroup extends InputBase {
  placeholder: string;

  constructor(
    name: string,
    label: string,
    placeholder: string,
    options: [string, string][],
    selectedOption: keyof typeof UNITS
  ) {
    super({
      name,
      label,
      options,
      selectedOption,
      inputCategory: "inputGroup",
    });
    this.placeholder = placeholder;
  }
}

export type Input = InputGroup | InputBase;
export type Inputs = typeof inputs;
export type InputCategory = "inputSelect" | "inputGroup";

export const inputs = {
  liquid: () =>
    new InputGroup(
      "liquid",
      "Liquid Drank per day",
      "50",
      [
        ["ml", "ml"],
        ["l", "l"],
      ],
      "ml"
    ),
  time: () =>
    new InputGroup(
      "time",
      "Time spent",
      "5",
      [
        ["minutes", "minutes"],
        ["hours", "hours"],
      ],
      "minutes"
    ),
  frequency: () =>
    new InputBase({
      name: "frequency",
      label: "Time spent",
      options: [
        [1, "1 time per week"],
        [2, "2 times per week"],
        [3, "3 times per week"],
        [4, "4 times per week"],
        [5, "5 times per week"],
        [6, "6 times per week"],
        [7, "Every day ðŸš€"],
      ],
      selectedOption: 3,
    }),
  cigarettes: () =>
    new InputGroup(
      "generic",
      "Cigarettes you want to avoid per day",
      "1",
      [["generic", "cigarettes"]],
      "generic"
    ),
  smartphone: () =>
    new InputGroup(
      "generic",
      "Hours to avoid per day",
      "1",
      [["generic", "hours"]],
      "generic"
    ),
};
