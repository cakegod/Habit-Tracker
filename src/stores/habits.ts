import { atom } from "nanostores";
import type { Habit } from "@data/habits";
import type { HabitData } from "@components/home/ModalForm";

export type HabitStateData = Habit & HabitData;
export const habits = atom<HabitStateData[]>([]);

// TODO: Refactor into two separated functions
export const addHabit = (habit: HabitStateData) => {
  const isPresent = !!habits.get().find((h) => h.id === habit.id);

  isPresent
    ? // Updating the object with the argument if is present in the array
      habits.set(habits.get().map((h) => (h.id === habit.id ? habit : h)))
    : habits.set([...habits.get(), habit]);

  console.log(habits.get());
};

export const deleteHabit = (habit: HabitStateData) => {
  habits.set(habits.get().filter((h) => h.id !== habit.id));

  console.log(habits);
};

export const currentHabit = atom<HabitStateData>({
  name: "",
  icon: "",
  id: "",
  forms: [],
  time: {
    value: 0,
    type: "minutes",
  },
  frequency: {
    value: 0,
  },
  liquid: {
    value: 0,
    type: "ml",
  },
});

export const isModalOpen = atom(false);

export const toggleModal = (habit?: HabitStateData) => {
  isModalOpen.set(!isModalOpen.get());
  if (habit) {
    currentHabit.set(habit);
  }
  console.table(habits);
};
