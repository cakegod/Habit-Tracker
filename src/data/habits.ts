import { inputs as input } from "./inputs";
import { nanoid } from "nanoid";

export type HabitsNames = (typeof habitsData)[number]["name"];

function createHabit<Name extends string>(
  name: Name,
  icon: string,
  inputs: ReadonlyArray<(typeof input)[keyof typeof input]>,
  avoid: boolean = false
) {
  return { name, icon, inputs, avoid, id: nanoid() };
}

export type HabitData = ReturnType<typeof createHabit>;
export type Habit = (typeof habitsData)[number];

// TODO: add more habits
export const habitsData = [
  createHabit("Meditate", "🧘‍♀️", [input.time, input.frequency]),
  createHabit("Exercise", "🏋️‍♂️", [input.time, input.frequency]),
  createHabit("Read", "📖", [input.time, input.frequency]),
  createHabit("Write", "📝", [input.time, input.frequency]),
  createHabit("Drink Water", "💧", [input.liquid]),
  createHabit("Code", "👨‍💻", [input.time, input.frequency]),
  createHabit("Smoke", "🚬", [input.cigarettes], true),
  createHabit("Smartphone", "📱", [input.smartphone], true),
  createHabit("Learn Language", "🌎", [input.time, input.frequency]),
].sort((a) => (a.avoid ? 1 : -1)) satisfies Readonly<HabitData[]>;
