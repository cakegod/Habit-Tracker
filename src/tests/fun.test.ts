import { createHabit } from "@data/habits";
import { inputs } from "@data/inputs";
import { describe, expect, test } from "vitest";

describe("habits", () => {
  test("test", () => {
    const habit = createHabit("Meditate", "🧘‍♀️", [
      inputs.time,
      inputs.frequency,
    ]);
    expect(habit.rawData()).toMatchSnapshot({
      id: expect.any(String),
    });

		expect(habit.inputNames).toMatchSnapshot()

		expect(habit.unit).toMatchSnapshot()

  });
});
