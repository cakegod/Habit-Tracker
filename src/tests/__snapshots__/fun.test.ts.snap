// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`habits > test 1`] = `
{
  "avoid": false,
  "icon": "🧘‍♀️",
  "id": Any<String>,
  "inputs": [
    InputGroup {
      "inputCategory": "inputGroup",
      "label": "Time spent",
      "name": "time",
      "options": [
        [
          "minutes",
          "minutes",
        ],
        [
          "hours",
          "hours",
        ],
      ],
      "placeholder": "5",
      "selectedOption": "minutes",
      "type": "number",
      "value": "",
    },
    InputBase {
      "inputCategory": "inputSelect",
      "label": "Time spent",
      "name": "frequency",
      "options": [
        [
          1,
          "1 time per week",
        ],
        [
          2,
          "2 times per week",
        ],
        [
          3,
          "3 times per week",
        ],
        [
          4,
          "4 times per week",
        ],
        [
          5,
          "5 times per week",
        ],
        [
          6,
          "6 times per week",
        ],
        [
          7,
          "Every day 🚀",
        ],
      ],
      "selectedOption": 3,
      "type": "number",
      "value": "",
    },
  ],
  "name": "Meditate",
}
`;

exports[`habits > test 2`] = `
{
  "frequency": InputBase {
    "inputCategory": "inputSelect",
    "label": "Time spent",
    "name": "frequency",
    "options": [
      [
        1,
        "1 time per week",
      ],
      [
        2,
        "2 times per week",
      ],
      [
        3,
        "3 times per week",
      ],
      [
        4,
        "4 times per week",
      ],
      [
        5,
        "5 times per week",
      ],
      [
        6,
        "6 times per week",
      ],
      [
        7,
        "Every day 🚀",
      ],
    ],
    "selectedOption": 3,
    "type": "number",
    "value": "",
  },
  "time": InputGroup {
    "inputCategory": "inputGroup",
    "label": "Time spent",
    "name": "time",
    "options": [
      [
        "minutes",
        "minutes",
      ],
      [
        "hours",
        "hours",
      ],
    ],
    "placeholder": "5",
    "selectedOption": "minutes",
    "type": "number",
    "value": "",
  },
}
`;

exports[`habits > test 3`] = `"minutes"`;
